server/keycloak.ts

```diff
@@ -373,6 +402,65 @@ class Keycloak {
       throw error;
     }
   }
+
+  /**
+   * Migrate user's old keycloak id to new keycloak id and update roles
+   *
+   * @param keycloakId
+   * @param email
+   * @param username
+   */
+  async migrateUser(keycloakId: string, email: string, username: string): Promise<string[] | null> {
+    const migrationStatus = await dbClient.db[collections.USER_MIGRATION].findOne({
+      'email ilike': email,
+      'username ilike': username.includes('@bceid') ? `${username.split('@')[0]}@bceid%` : username,
+    });
+
+    if (!migrationStatus || migrationStatus.status !== 'pending') {
+      return null;
+    }
+
+    const undefinedRoles = migrationStatus.roles.filter((role) => !this.roleIdMap[role]);
+
+    if (undefinedRoles.length > 0) {
+      const msg = `${undefinedRoles.join(', ')} not defined`;
+      logger.error(msg, { context: 'kc-user-migration' });
+      throw Error(msg);
+    }
+
+    try {
+      await this.setUserRoles(keycloakId, migrationStatus.roles);
+    } catch (e) {
+      logger.error(`failed to update user(${keycloakId})'s role`, { context: 'kc-user-migration' });
+      throw e;
+    }
+
+    const [user] = await dbClient.db[collections.USERS].findDoc({
```

Comment:
Can I suggest instead storing the original keycloak ID in the table and using that instead of username match on the users table?

Reply:
I don't catch what you mean. I want to make sure that this logic doesn't require access to both Keycloaks.

1. Run database migration to create `user_migration` table with OCIO KEYCLOAK.
2. Switch to MoH KEYCLOAK
3. User log in with the MoH KEYCLOAK
4. Keycloak creates an account for the user
5. HCAP updates the user's roles on MoH KEYCLOAK.

Reply:
No need for this, thank you!